src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExpressionEvaluator
+{
+    internal sealed class EEMetadataReferenceResolver : MetadataReferenceResolver
+    {
+        private readonly AssemblyIdentityComparer _identityComparer;
+        private readonly Dictionary<string, ImmutableArray<(AssemblyIdentity, MetadataReference)>> _referencesByIdentity;
+
+#if DEBUG
+        internal readonly Dictionary<AssemblyIdentity, (AssemblyIdentity Identity, int Count)> Requests = new Dictionary<AssemblyIdentity, (AssemblyIdentity Identity, int Count)>();
+#endif
+
+        internal EEMetadataReferenceResolver(AssemblyIdentityComparer identityComparer, Dictionary<string, ImmutableArray<(AssemblyIdentity, MetadataReference)>> referencesByIdentity)
+        {
+            _identityComparer = identityComparer;
+            _referencesByIdentity = referencesByIdentity;
+        }
+
+        public override bool ResolveMissingAssemblies => true;
+
+        public override PortableExecutableReference ResolveMissingAssembly(MetadataReference definition, AssemblyIdentity referenceIdentity)
+        {
+            ImmutableArray<(AssemblyIdentity, MetadataReference)> references;
+            (AssemblyIdentity, MetadataReference) result = default;
```

Comment:
>AssemblyIdentity, MetadataReference [](start = 13, length = 35)

Perhaps we could name these so that we don't need to wonder what `Item1` and `Item2` are below